src/layout.ts

```diff
@@ -48,17 +48,43 @@ export function layoutPage(content: Paragraph[], box: Box, fonts: Font[]): Frame
 function layoutParagraph(content: Paragraph, box: Box, fonts: Font[]): Frame {
   const maxWidth = box.width;
   const maxHeight = box.height;
-  const { text, ...attrs } = content;
-  const segments = extractTextSegments(text, attrs, fonts);
-  return layoutRow(segments, { ...box, width: maxWidth, height: maxHeight });
+  const innerBox = { x: 0, y: 0, width: maxWidth, height: maxHeight };
+  const rows = content.text && layoutText(content, innerBox, fonts);
+  return {
+    type: 'paragraph',
+    ...box,
+    width: rows?.size?.width ?? 0,
+    height: rows?.size?.height ?? 0,
+    ...(rows?.rows?.length ? { children: rows.rows } : undefined),
+  };
+}
+
+function layoutText(content: Paragraph, box: Box, fonts: Font[]) {
+  const { text } = content;
+  const segments = extractTextSegments(text, content, fonts);
+  const rows = [];
+  let remainingSegments = segments;
+  const remainingSpace = { ...box };
+  const size = { width: 0, height: 0 };
+  while (remainingSegments?.length) {
+    const { row, remainder } = layoutRow(remainingSegments, remainingSpace);
+    rows.push(row);
+    remainingSegments = remainder;
+    remainingSpace.height -= row.height;
+    remainingSpace.y += row.height;
+    size.width = Math.max(size.width, row.width);
+    size.height = row.y + row.height - box.y;
```

Comment:
Can `size.height = row.y + row.height - box.y;` be out of the loop and done only once for the last row.
Is box.y always zero here?